Fix bug when multiple files fail file checks

wthrowe opened the pull request:
If two files fail the same check then the pretty_grep function will be
supplied multiple filenames and needs to process them all individually.

staged_grep is only ever called with one file, so changes there are
unnecessary.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Fix bug when multiple files fail file checks

If two files fail the same check then the pretty_grep function will be
supplied multiple filenames and needs to process them all individually.

staged_grep is only ever called with one file, so changes there are
unnecessary.

nilsdeppe commented:
@hrueter could you please review this since you seem to actually be qualified to (unlike @kidder and I)?

wthrowe commented:
I adjusted the option parsing a bit.  Since no one had commented on the code I squashed immediately.

moxcodes commented on tools/FileTestDefs.sh:
```diff
@@ -38,9 +38,34 @@ staged_grep() {
-        GREP_COLOR='1;37;41' grep -n $color_option "${@:1:$(($#-1))}"
+    local -a args
+    local file file_prefix
+    while [ $# -gt 0 ] ; do
```
this would be easier to understand if you added a comment explaining that this loop accumulates all of the dash-preceded flags, followed by the grep pattern, and that following the loop the only arguments remaining in $@ will be file names.

moxcodes commented on tools/FileTestDefs.sh:
```diff
@@ -38,9 +38,34 @@ staged_grep() {
 # Utility function for reporters that enables lots of decorators in grep
 # Works like staged_grep
```
it would be helpful to have a comment here indicating the expected syntax for the function, specifically `pretty_grep [-flags...] [--] pattern filenames...`

moxcodes commented on tools/FileTestDefs.sh:
```diff
@@ -38,9 +38,34 @@ staged_grep() {
-    echo -n -e "\033[0;35m${@: -1}\033[0m:"
-    git show ":./${@: -1}" | \
-        GREP_COLOR='1;37;41' grep -n $color_option "${@:1:$(($#-1))}"
+    local -a args
```
args would be easier to understand if called something along the lines of `grep_options` or `grep_args`

moxcodes reviewed (commented):
Group code review with @markscheel and @fmahebert

wthrowe commented:
Added some comments and rebased to pick up the Travis changes.

hrueter commented on tools/FileTestDefs.sh:
```diff
@@ -35,12 +35,42 @@ staged_grep() {
+    # non_file_args, leaving the filenames in $@.
+    while [ $# -gt 0 ] ; do
+        case "$1" in
+            --)
```
What are two dashes in a grep command doing? I could not find it in the grep manual.

wthrowe replied:
Yeah, I was a bit surprised they did not document it explicitly.  It means to stop processing switches so you can pass other options (usually files, in this case a pattern) that begin with a `-` without them being interpreted as switches.  It is a standard GNU-style option parsing thing, so I guess since they didn't implement it they left it out (intentionally or accidentally).

nilsdeppe replied:
Would it be maybe be good to add a comment on what it does if it's not in the grep manual?

hrueter replied:
Ah thanks, I remember now where I have used this before. 

@wthrowe However, your implementation does not work when `--` is used after the pattern.

wthrowe replied:
I didn't think of that.  Handling that case sounds annoying.  Would you be OK with a comment telling people not do to that instead?  (Any "--" case people want can be trivially rewritten into the case that is already handled.)

wthrowe committed:
Stop complaining about acceptable long lines

Output coloring is now enabled unconditionally, so piping the report
output through grep won't break it anymore.

Fixes #1371.

nilsdeppe reviewed (dismissed):
Thanks for these fixes @wthrowe!

@hrueter could you please look at the fixup and the new commit whenever you get a chance? Thanks!

hrueter reviewed (dismissed):
I am happy with it.

kidder commented:
@wthrowe please squash
